complex/ft_test.c

```diff
@@ -269,6 +269,110 @@ static int ft_run_latency(void)
 	return 0;
 }
 
+static int ft_bw(void)
+{
+	int ret, i;
+	if (listen_sock < 0) {
+		for (i = 0; i < ft.xfer_iter; i++) {
+                        ret = ft_send_msg();
+			if (ret)
+				return ret;
+		}
+                while (ft_tx.credits < ft_tx.max_credits) {
+                        ret = ft_comp_tx();
+                        if (ret)
+                                return ret;
+                }
+                ret = ft_recv_msg();
+                if (ret)
+			return ret;
+	} else {
+		int count=0;
+                do {
+			ret = ft_post_recv_bufs();
+			if (ret)
+				return ret;
+			ret = ft_comp_rx();
+			if (ret)
+				return ret;
+			count+=ft_rx.credits;
+                } while (count < ft.xfer_iter);
+                ret = ft_send_msg();
+                if (ret)
+			return ret;
+                while (ft_tx.credits < ft_tx.max_credits) {
+			ret = ft_comp_tx();
+			if (ret)
+				return ret;
+                }
+	}
+	return 0;
+}
+
+static int ft_bw_dgram(void)
+{
+	int ret, i;
+
+	if (listen_sock < 0) {
+		for (i = 0; i < ft.xfer_iter; i++) {
+			ret = ft_send_dgram();
+			if (ret)
+				return ret;
+		}
+		ret = ft_send_dgram_done();
+		if(ret)
+			return ret;
+		int credits = 0;
+		while(credits)
+		  {
+			  ret = ft_post_recv_bufs();
+			  if (ret)
+				  return ret;
+			  ret = ft_comp_rx();
+			  if (ret)
+				  return ret;
+			  credits=ft_rx.credits;
+		  }
+	} else {
+          ret = ft_recv_dgram_flood();
+          if (ret)
+                return ret;
+          ret = ft_send_dgram();
+                return ret;
```

Comment:
nit: there appears to be a missing if statement here, but it doesn't affect the code